File: src/main/java/no/jstien/bikeapi/station/AvailabilityPredictor.java
package no.jstien.bikeapi.station;

import no.jstien.bikeapi.tsdb.read.*;
import no.jstien.bikeapi.utils.AnalogueDateFinder;
import no.jstien.bikeapi.utils.CalendarUtils;

import java.time.ZonedDateTime;
import java.util.*;

public class AvailabilityPredictor {
    private static final int MINUTES_PER_DAY = 60 * 24;
    private static final int DEFAULT_DOWNSAMPLE_MINUTES = 15;

    private StationTSDBReader tsdbReader;
    private AnalogueDateFinder dateFinder;
    private int downsampleMinutes = DEFAULT_DOWNSAMPLE_MINUTES;

    public AvailabilityPredictor(StationTSDBReader tsdbReader, AnalogueDateFinder analogueDateFinder) {
        this.tsdbReader = tsdbReader;
        this.dateFinder = analogueDateFinder;
    }

    public void setDownsampleMinutes(int minutes) {
        this.downsampleMinutes = minutes;
    }

    public StationHistory predictForStation(int stationId, Calendar predictionDate) {
        // TODO: This is quite nasty, and I'm honestly surprised if it turns out to not be
        // buggy AF. Anyways, it's an OK POC I guess.
        List<Calendar> analogues = dateFinder.findAnaloguesForDay(predictionDate);

        final int numElems = MINUTES_PER_DAY / downsampleMinutes;
        int[] bikesCount = new int[numElems];
        double[] bikes = new double[numElems];
        int[] locksCount = new int[numElems];
        double[] locks = new double[numElems];

        analogues.forEach(day -> {
            GregorianCalendar startDate = CalendarUtils.startOfDay(day);
            final long startTime = startDate.getTimeInMillis() / 1000L;

            ZonedDateTime from = startDate.toZonedDateTime();
            ZonedDateTime to = CalendarUtils.endOfDay(day).toZonedDateTime();
            RequestFactory requestFactory = new RequestFactory(from, to, stationId);
            requestFactory.setDownsampleMinutes(downsampleMinutes);

            Map<Integer, StationHistory> history = tsdbReader.queryStations(requestFactory);
            history.values().stream()
                .forEach(h -> {
                    processTimeSerie(h.getFreeBikes(), startTime, bikes, bikesCount);
                    processTimeSerie(h.getFreeLocks(), startTime, locks, locksCount);
                });
        });

        // Cool, let's try to create a StationHistory from it
        reduceHistogram(bikes, bikesCount);
        reduceHistogram(locks, locksCount);

        final long timeZero = CalendarUtils.startOfDay(predictionDate).getTimeInMillis() / 1000L;
        TimeSerie bikeSerie = toTimeSerie(bikes, timeZero);
        TimeSerie lockSerie = toTimeSerie(locks, timeZero);
        return new StationHistory(stationId, bikeSerie, lockSerie);
    }

    private void processTimeSerie(TimeSerie timeSerie, long startTime, double[] dest, int[] counts) {
        timeSerie.getDataPoints().forEach(dp -> {
            final long timestamp = dp.getTimestamp();
            final long index = (timestamp - startTime) / (downsampleMinutes * 60);
            dest[(int)index] += dp.getValue();
            counts[(int)index]++;
        });
    }

    private void reduceHistogram(double[] vals, int[] counts) {
        for (int i=0; i<vals.length; i++) {
            if (counts[i] != 0) {
                vals[i] /= (double)counts[i];
            }
        }
    }

    private TimeSerie toTimeSerie(double[] vals, long firstTimeStamp) {
        TimeSerie ts = new TimeSerie();

        final long deltaTime = (downsampleMinutes * 60);
        long time = firstTimeStamp;
        for (int i=0; i<vals.length; i++) {
            ts.addDataPoint(new DataPoint(time, vals[i]));
            time += deltaTime;
        }

        return ts;
    }

}

File: src/main/java/no/jstien/bikeapi/station/AvailabilityPredictor.kt
package no.jstien.bikeapi.station

import no.jstien.bikeapi.tsdb.read.*
import no.jstien.bikeapi.utils.AnalogueDateFinder
import no.jstien.bikeapi.utils.CalendarUtils
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger

import java.time.ZonedDateTime
import java.util.Calendar
import java.util.GregorianCalendar

class AvailabilityPredictor(private val tsdbReader: StationTSDBReader, private val dateFinder: AnalogueDateFinder) {
    private var downsampleMinutes = DEFAULT_DOWNSAMPLE_MINUTES

    fun setDownsampleMinutes(minutes: Int) {
        this.downsampleMinutes = minutes
    }

    fun predictForStation(stationId: Int, predictionDate: Calendar): StationHistory {
        // TODO: This is quite nasty, and I'm honestly surprised if it turns out to not be
        // buggy AF. Anyways, it's an OK POC I guess.
        val analogues = dateFinder.findAnaloguesForDay(predictionDate)

        val numElems = MINUTES_PER_DAY / downsampleMinutes
        val bikesCount = IntArray(numElems)
        val bikes = DoubleArray(numElems)
        val locksCount = IntArray(numElems)
        val locks = DoubleArray(numElems)

        analogues.forEach { day ->
            val startDate = CalendarUtils.startOfDay(day)
            val from = startDate.toZonedDateTime()
            val to = CalendarUtils.endOfDay(day).toZonedDateTime()
            val requestFactory = RequestFactory(from, to, stationId)
            requestFactory.setDownsampleMinutes(downsampleMinutes)

            val startTime = startDate.timeInMillis / 1000L
            val history = tsdbReader.queryStations(requestFactory)
            history.values.stream()
                    .forEach { h ->
                        processTimeSerie(h.freeBikes!!, startTime, bikes, bikesCount)
                        processTimeSerie(h.freeLocks!!, startTime, locks, locksCount)
                    }
        }

        // Cool, let's try to create a StationHistory from it
        reduceHistogram(bikes, bikesCount)
        reduceHistogram(locks, locksCount)

        val timeZero = CalendarUtils.startOfDay(predictionDate).timeInMillis / 1000L
        val bikeSerie = toTimeSerie(bikes, timeZero)
        val lockSerie = toTimeSerie(locks, timeZero)
        return StationHistory(stationId, bikeSerie, lockSerie)
    }

    private fun processTimeSerie(timeSerie: TimeSerie, startTime: Long, dest: DoubleArray, counts: IntArray) {
        timeSerie.dataPoints.forEach { dp ->
            val timestamp = dp.timestamp
            val index = (timestamp - startTime) / (downsampleMinutes * 60)
            dest[index.toInt()] += dp.value
            counts[index.toInt()]++
        }
    }

    private fun reduceHistogram(vals: DoubleArray, counts: IntArray) {
        for (i in vals.indices) {
            if (counts[i] != 0) {
                vals[i] /= counts[i].toDouble()
            }
        }
    }

    private fun toTimeSerie(vals: DoubleArray, firstTimeStamp: Long): TimeSerie {
        val ts = TimeSerie()

        val deltaTime = (downsampleMinutes * 60).toLong()
        var time = firstTimeStamp
        for (i in vals.indices) {
            ts.addDataPoint(DataPoint(time, vals[i]))
            time += deltaTime
        }

        return ts
    }

    companion object {
        private val LOG = LogManager.getLogger()

        private val MINUTES_PER_DAY = 60 * 24
        private val DEFAULT_DOWNSAMPLE_MINUTES = 15
    }

}
